File: caster.ui/src/app/exercise/state/exercise.service.ts
/*
Crucible
Copyright 2020 Carnegie Mellon University.
NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR IMPLIED, AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF FITNESS FOR PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS OBTAINED FROM USE OF THE MATERIAL. CARNEGIE MELLON UNIVERSITY DOES NOT MAKE ANY WARRANTY OF ANY KIND WITH RESPECT TO FREEDOM FROM PATENT, TRADEMARK, OR COPYRIGHT INFRINGEMENT.
Released under a MIT (SEI)-style license, please see license.txt or contact permission@sei.cmu.edu for full terms.
[DISTRIBUTION STATEMENT A] This material has been approved for public release and unlimited distribution.  Please see Copyright notice for non-US Government use and distribution.
Carnegie Mellon® and CERT® are registered in the U.S. Patent and Trademark Office by Carnegie Mellon University.
DM20-0181
*/

import { ExerciseStore } from './exercise.store';
import { Injectable } from '@angular/core';
import { ExerciseUI, Breadcrumb, ExerciseObjectType, Tab } from './exercise.model';
import { DirectoryQuery } from 'src/app/directories';
import { ExerciseQuery } from './exercise.query';
import { Workspace, Directory, Exercise, ExercisesService, ModelFile } from 'src/app/generated/caster-api';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { WorkspaceQuery, WorkspaceService } from 'src/app/workspace/state';

@Injectable({
  providedIn: 'root'
})
export class ExerciseService {
  constructor(
    private exerciseStore: ExerciseStore,
    private directoryQuery: DirectoryQuery,
    private workspaceQuery: WorkspaceQuery,
    private exerciseQuery: ExerciseQuery,
    private exercisesService: ExercisesService
    ) {

  }

  loadExercises(): Observable<Exercise[]> {
    return this.exercisesService.getAllExercises().pipe(
      tap(exercises => {
        const exerciseUIs = this.exerciseQuery.ui.getAll();
        this.exerciseStore.set(exercises);
        exercises.forEach(exercise => {
          const exerciseUI = exerciseUIs.find(xUI => (xUI.id === exercise.id));
          if (exerciseUI) {
            this.exerciseStore.ui.upsert(exercise.id, exerciseUI);
          }
        });
      }),
    );
  }

  loadExercise(exerciseId: string): Observable<Exercise> {
    return this.exercisesService.getExercise(exerciseId).pipe(
      tap(exercise => {
        this.exerciseStore.upsert(exercise.id, exercise);
        this.exerciseStore.ui.upsert(exercise.id, this.exerciseQuery.ui.getEntity(exercise.id));
      }),
    );
  }

  createExercise(exercise: Exercise): Observable<Exercise> {
    return this.exercisesService.createExercise(exercise).pipe(
      tap(ex => {
        this.exerciseStore.add(ex);
        this.exerciseStore.ui.upsert(ex.id, this.exerciseQuery.ui.getEntity(ex.id));
      }),
    );
  }

  deleteExercise(exerciseId: string): Observable<any> {
    return this.exercisesService.deleteExercise(exerciseId).pipe(
      tap(() => {
        this.exerciseStore.remove(exerciseId);
        this.exerciseStore.ui.remove(exerciseId);
      }),
    );
  }

  updateExercise(exercise: Exercise): Observable<Exercise> {
    return this.exercisesService.editExercise(exercise.id, exercise).pipe(
      tap(ex => {
        this.exerciseStore.upsert(ex.id, ex);
        this.exerciseStore.ui.upsert(ex.id, this.exerciseQuery.ui.getEntity(ex.id));
      }),
    );
  }

  setSelectedTab(index: number) {
    const exercise = this.exerciseQuery.getActive();
    if (exercise) {
      let exUi = this.exerciseQuery.ui.getEntity(exercise.id);
      exUi = { ...exUi, selectedTab: index } as ExerciseUI;
      this.exerciseStore.ui.upsert(exUi.id, exUi);
    }
  }

  closeTab(id: string) {
    const exercise = this.exerciseQuery.getActive();
    if (exercise) {
      const exerciseUI = this.exerciseQuery.ui.getEntity(exercise.id);
      const tabs = new Array<Tab>().concat(exerciseUI.openTabs);
      const index = exerciseUI.openTabs.findIndex(e => e.id === id);
      tabs.splice(index, 1);
      const selected = (exerciseUI.selectedTab >= tabs.length) ? tabs.length - 1 : exerciseUI.selectedTab;
      const exUi = { ...exerciseUI, openTabs: tabs, selectedTab:  selected};
      this.exerciseStore.ui.upsert(exUi.id, exUi);
    }
  }


  openTab(obj: ModelFile | Workspace, objType: ExerciseObjectType) {
    const exercise = this.exerciseQuery.getActive();
    if (exercise) {
      const exerciseUI = this.exerciseQuery.ui.getEntity(exercise.id);
      const tabIndex = exerciseUI.openTabs.findIndex(t => t.id === obj.id);
      if (tabIndex > -1) {
        // Tab already open so simply select it
        if (exerciseUI.selectedTab !== tabIndex) {
          const exUi = { ...exerciseUI, selectedTab: tabIndex } as ExerciseUI;
          this.exerciseStore.ui.upsert(exUi.id, exUi);
        }
      } else {
        // Tab is not open, add it
        let tabs = new Array<Tab>();
        tabs = tabs.concat(exerciseUI.openTabs);
        tabs.push(
          {
            id: obj.id,
            name: obj.name,
            type: objType,
            directoryId: obj.directoryId
          } as Tab);
        const exUi = { ...exerciseUI, openTabs: tabs, selectedTab: tabs.length - 1 } as ExerciseUI;
        this.exerciseStore.ui.upsert(exUi.id, exUi);
      }
    }
  }


  updateTabBreadcrumb(tabId: string, newBreadcrumb: Breadcrumb[]) {
    const exercise = this.exerciseQuery.getActive();
    let isChanged = false;
    if (exercise) {
      const exerciseUI = this.exerciseQuery.ui.getEntity(exercise.id);
      let updatedTabs = new Array<Tab>();
      exerciseUI.openTabs.forEach(tab => {
        if (tab.id === tabId) {
          if (!tab.breadcrumb) {
            isChanged = true;
          } else if (newBreadcrumb.length === tab.breadcrumb.length) {
            for (let i = 0; i < newBreadcrumb.length; i++) {
              if (newBreadcrumb[i].id !== tab.breadcrumb[i].id ||
                  newBreadcrumb[i].name !== tab.breadcrumb[i].name ||
                  newBreadcrumb[i].type !== tab.breadcrumb[i].type) {
                isChanged = true;
              }
            }
          }
          if (isChanged) {
            const newTab = {
              id: tab.id,
              name: tab.name,
              type: tab.type,
              directoryId: tab.directoryId,
              breadcrumb: newBreadcrumb
            };
            updatedTabs.push(newTab);
          }
        } else {
          updatedTabs.push(tab);
        }
      });
      if (isChanged) {
        const exUi = { ...exerciseUI, openTabs: updatedTabs } as ExerciseUI;
        this.exerciseStore.ui.upsert(exUi.id, exUi);
      }
    }
  }


  createBreadcrumb(exercise: Exercise, obj: ModelFile | Workspace | Tab , objType: ExerciseObjectType): Array<Breadcrumb> {
    const breadcrumb: Breadcrumb[] = [];
    breadcrumb.push({ name: obj.name, id: obj.id, type: objType } as Breadcrumb);
    if (objType === ExerciseObjectType.FILE && (obj as ModelFile).workspaceId !== null) {
      // When the file belongs to a workspace, add the workspace to the breadcrumb
      const workspace = this.workspaceQuery.getEntity((obj as ModelFile).workspaceId);
      if (workspace) {
        breadcrumb.unshift({ name: workspace.name, id: workspace.id, type: ExerciseObjectType.WORKSPACE } as Breadcrumb);
      }
    }

    let currentDir: Directory = this.directoryQuery.getEntity(obj.directoryId);
    while (currentDir) {
      breadcrumb.unshift({ name: currentDir.name, id: currentDir.id, type: ExerciseObjectType.DIRECTORY } as Breadcrumb);
      currentDir = this.directoryQuery.getEntity(currentDir.parentId);
    }
    breadcrumb.unshift({ name: exercise.name, id: exercise.id, type: ExerciseObjectType.EXERCISE } as Breadcrumb);

    return breadcrumb;
  }

  setRightSidebarOpen(exerciseId: string, open: boolean) {
    this.exerciseStore.ui.upsert(exerciseId, {rightSidebarOpen: open});
  }

  setRightSidebarView(exerciseId: string, view: string) {
    this.exerciseStore.ui.upsert(exerciseId, {rightSidebarView: view});
  }

  setRightSidebarWidth(exerciseId: string, width: number) {
    this.exerciseStore.ui.upsert(exerciseId, {rightSidebarWidth: width});
  }

  setLeftSidebarOpen(exerciseId: string, open: boolean) {
    this.exerciseStore.ui.upsert(exerciseId, {leftSidebarOpen: open});
  }

  setLeftSidebarWidth(exerciseId: string, width: number) {
    this.exerciseStore.ui.upsert(exerciseId, {leftSidebarWidth: width});
  }
}

File: player.api/S3.Player.Api/Services/ExerciseService.cs
/*
Crucible
Copyright 2020 Carnegie Mellon University.
NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR IMPLIED, AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF FITNESS FOR PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS OBTAINED FROM USE OF THE MATERIAL. CARNEGIE MELLON UNIVERSITY DOES NOT MAKE ANY WARRANTY OF ANY KIND WITH RESPECT TO FREEDOM FROM PATENT, TRADEMARK, OR COPYRIGHT INFRINGEMENT.
Released under a MIT (SEI)-style license, please see license.txt or contact permission@sei.cmu.edu for full terms.
[DISTRIBUTION STATEMENT A] This material has been approved for public release and unlimited distribution.  Please see Copyright notice for non-US Government use and distribution.
Carnegie Mellon(R) and CERT(R) are registered in the U.S. Patent and Trademark Office by Carnegie Mellon University.
DM20-0181
*/

using AutoMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.EntityFrameworkCore;
using S3.Player.Api.Data.Data;
using S3.Player.Api.Data.Data.Models;
using S3.Player.Api.Extensions;
using S3.Player.Api.Infrastructure.Authorization;
using S3.Player.Api.Infrastructure.Exceptions;
using S3.Player.Api.ViewModels;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Security.Principal;
using System.Threading;
using System.Threading.Tasks;

namespace S3.Player.Api.Services
{
    public interface IExerciseService
    {
        Task<IEnumerable<ViewModels.Exercise>> GetAsync(CancellationToken ct);
        Task<ViewModels.Exercise> GetAsync(Guid id, CancellationToken ct);
        Task<IEnumerable<ViewModels.Exercise>> GetByUserIdAsync(Guid userId, CancellationToken ct);
        Task<ViewModels.Exercise> CreateAsync(ViewModels.Exercise exercise, CancellationToken ct);
        Task<Exercise> CloneAsync(Guid id, CancellationToken ct);
        Task<ViewModels.Exercise> UpdateAsync(Guid id, ViewModels.Exercise exercise, CancellationToken ct);
        Task<bool> DeleteAsync(Guid id, CancellationToken ct);
    }

    public class ExerciseService : IExerciseService
    {
        private readonly PlayerContext _context;
        private readonly IAuthorizationService _authorizationService;
        private readonly ClaimsPrincipal _user;
        private readonly IMapper _mapper;
        private IUserClaimsService _claimsService;

        public ExerciseService(PlayerContext context, IAuthorizationService authorizationService, IPrincipal user, IMapper mapper, IUserClaimsService claimsService)
        {
            _context = context;
            _authorizationService = authorizationService;
            _user = user as ClaimsPrincipal;
            _mapper = mapper;
            _claimsService = claimsService;
        }

        public async Task<IEnumerable<ViewModels.Exercise>> GetAsync(CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new FullRightsRequirement())).Succeeded)
                throw new ForbiddenException();

            var items = await _context.Exercises
                .ToListAsync(ct);         
            
            return _mapper.Map<IEnumerable<Exercise>>(items);
        }

        public async Task<ViewModels.Exercise> GetAsync(Guid id, CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new FullRightsRequirement())).Succeeded &&
                !(await _authorizationService.AuthorizeAsync(_user, null, new ExerciseMemberRequirement(id))).Succeeded)
                throw new ForbiddenException();

            var item = await _context.Exercises
                .SingleOrDefaultAsync(o => o.Id == id, ct);

            return _mapper.Map<Exercise>(item);
        }

        public async Task<IEnumerable<ViewModels.Exercise>> GetByUserIdAsync(Guid userId, CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new SameUserRequirement(userId))).Succeeded)
                throw new ForbiddenException();

            var user = await _context.Users
                .Include(u => u.ExerciseMemberships)
                    .ThenInclude(em => em.Exercise)                
                .Where(u => u.Id == userId)
                .SingleOrDefaultAsync(ct);

            if (user == null)
                throw new EntityNotFoundException<User>();

            var exercises = user.ExerciseMemberships.Select(x => x.Exercise);

            return _mapper.Map<IEnumerable<ViewModels.Exercise>>(exercises);
        }
        
        public async Task<ViewModels.Exercise> CreateAsync(ViewModels.Exercise exercise, CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new ExerciseCreationRequirement())).Succeeded)
                throw new ForbiddenException();

            var exerciseEntity = Mapper.Map<ExerciseEntity>(exercise);

            var exerciseAdminPermission = await _context.Permissions
                .Where(p => p.Key == PlayerClaimTypes.ExerciseAdmin.ToString())
                .FirstOrDefaultAsync(ct);

            if (exerciseAdminPermission == null)
                throw new EntityNotFoundException<Permission>($"{PlayerClaimTypes.ExerciseAdmin.ToString()} Permission not found.");

            var userId = _user.GetId();

            // Create an Admin team with the caller as a member
            var teamEntity = new TeamEntity() { Name = "Admin" };
            teamEntity.Permissions.Add(new TeamPermissionEntity() { Permission = exerciseAdminPermission });

            var exerciseMembershipEntity = new ExerciseMembershipEntity { Exercise = exerciseEntity, UserId = userId };            
            exerciseEntity.Teams.Add(teamEntity);
            exerciseEntity.Memberships.Add(exerciseMembershipEntity);

            _context.Exercises.Add(exerciseEntity);
            await _context.SaveChangesAsync(ct);

            var teamMembershipEntity = new TeamMembershipEntity { Team = teamEntity, UserId = userId, ExerciseMembership = exerciseMembershipEntity };
            exerciseMembershipEntity.PrimaryTeamMembership = teamMembershipEntity;
            _context.TeamMemberships.Add(teamMembershipEntity);
            _context.ExerciseMemberships.Update(exerciseMembershipEntity);
            await _context.SaveChangesAsync(ct);

            return await GetAsync(exerciseEntity.Id, ct);
        }

        public async Task<Exercise> CloneAsync(Guid idToBeCloned, CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new ExerciseCreationRequirement())).Succeeded)
                throw new ForbiddenException();

            var exercise = await _context.Exercises
                .Include(o => o.Teams)
                    .ThenInclude(o => o.Applications)
                .Include(o => o.Teams)
                    .ThenInclude(o => o.Permissions)
                .Include(o => o.Applications)
                .SingleOrDefaultAsync(o => o.Id == idToBeCloned, ct);

            var newExercise = exercise.Clone();
            newExercise.Name = $"Clone of {newExercise.Name}";

            //copy exercise applications
            foreach (var application in exercise.Applications)
            {
                var newApplication = application.Clone();
                newExercise.Applications.Add(newApplication);
            }

            //copy teams
            foreach (var team in exercise.Teams)
            {
                var newTeam = team.Clone();

                //copy team applications
                foreach (var applicationInstance in team.Applications)
                {
                    var newApplicationInstance = applicationInstance.Clone();
               
                    var application = exercise.Applications.FirstOrDefault(o => o.Id == applicationInstance.ApplicationId);
                    var newApplication = newExercise.Applications.FirstOrDefault(o => application != null && o.Name == application.Name);

                    newApplicationInstance.Application = newApplication;

                    newTeam.Applications.Add(newApplicationInstance);
                }

                //copy team permissions
                foreach (var permission in team.Permissions)
                {
                    var newPermission = new TeamPermissionEntity(newTeam.Id, permission.PermissionId);
                    newTeam.Permissions.Add(newPermission);
                }

                newExercise.Teams.Add(newTeam);
            }

            _context.Add(newExercise);
            await _context.SaveChangesAsync(ct);

            return _mapper.Map<Exercise>(newExercise);
        }

        public async Task<ViewModels.Exercise> UpdateAsync(Guid id, ViewModels.Exercise exercise, CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new ExerciseAdminRequirement(id))).Succeeded)
                throw new ForbiddenException();

            var exerciseToUpdate = await _context.Exercises.SingleOrDefaultAsync(v => v.Id == id, ct);

            if (exerciseToUpdate == null)
                throw new EntityNotFoundException<Exercise>();

            Mapper.Map(exercise, exerciseToUpdate);

            _context.Exercises.Update(exerciseToUpdate);
            await _context.SaveChangesAsync(ct);

            return _mapper.Map(exerciseToUpdate, exercise);
        }

        public async Task<bool> DeleteAsync(Guid id, CancellationToken ct)
        {
            if (!(await _authorizationService.AuthorizeAsync(_user, null, new ExerciseAdminRequirement(id))).Succeeded)
                throw new ForbiddenException();

            var exerciseToDelete = await _context.Exercises.SingleOrDefaultAsync(v => v.Id == id, ct);

            if (exerciseToDelete == null)
                throw new EntityNotFoundException<Exercise>();

            _context.Exercises.Remove(exerciseToDelete);
            await _context.SaveChangesAsync(ct);

            return true;
        }
    }
}

